### PR TITLE
[MooreToCore] Add always_comb and always_latch lowering support

### PR DESCRIPTION
**Pull request by maerhart**
* Make sure ALL blocks without predecessors are removed properly. Note: `use_empty` is not up-to-date when erasing blocks with the rewriter as it only marks them as replaced.
* Also remove unreachable blocks within `func.func`
* Add lowering support for `always_comb` and `always_latch` procedures

### CODE REVIEW

**Review by hailongSun2000 — APPROVED** [review]
Wow :astonished:! So nice :rocket:!

File: lib/Conversion/MooreToCore/MooreToCore.cpp [hunk 1]
```diff
@@ -104,27 +105,31 @@ static hw::ModulePortInfo getModulePortInfo(const TypeConverter &typeConverter,
 }
 
 /// Erase all dead blocks in a region.
-static void eraseDeadBlocks(PatternRewriter &rewriter, Region &region) {
+static void eraseDeadBlocks(Region &region) {
```
Comment (uenoku) [inline lib/Conversion/MooreToCore/MooreToCore.cpp:108]: Can we use [mlir::eraseUnreachableBlocks](https://mlir.llvm.org/doxygen/namespacemlir.html#a6ab9585c8607c12142232e0f58ebcc8b)?

File: lib/Conversion/MooreToCore/MooreToCore.cpp [hunk 2]
```diff
@@ -23,6 +23,7 @@
 #include "mlir/Pass/Pass.h"
 #include "mlir/Transforms/DialectConversion.h"
 #include "llvm/ADT/TypeSwitch.h"
+#include <mlir/IR/Iterators.h>
```
[the remark this answers is not in the document]
Comment (maerhart) [reply lib/Conversion/MooreToCore/MooreToCore.cpp:26]: Interesting that clangd decided to add this with angle brackets 🤔

[the remark this answers is not in the document]
Comment (uenoku) [reply lib/Conversion/MooreToCore/MooreToCore.cpp:26]: Yeah I also disabled header insertion with `-header-insertion=never` in clangd setting 😃

File: lib/Conversion/MooreToCore/MooreToCore.cpp [hunk 3]
```diff
@@ -1388,6 +1389,9 @@ void MooreToCorePass::runOnOperation() {
   MLIRContext &context = getContext();
   ModuleOp module = getOperation();
 
+  IRRewriter rewriter(module);
+  (void)mlir::eraseUnreachableBlocks(rewriter, module->getRegions());
+
```
Comment (fabianschuiki) [inline lib/Conversion/MooreToCore/MooreToCore.cpp:1394]: Very neat!

**Review by fabianschuiki — COMMENTED** [review]
Awesome! 🥳

**Pull request merged**